lms/views/helpers/_via.py

```diff
@@ -1,8 +1,82 @@
 """Via-related view helpers."""
+import re
 from urllib.parse import parse_qsl, urlencode, urlparse
 
 __all__ = ["via_url"]
 
+# pylint: disable=too-few-public-methods
+
+
+class _ViaDoc:
+    """A doc we want to proxy with content type."""
+
+    GOOGLE_DRIVE_REGEX = re.compile(
+        r"^https://drive.google.com/uc\?id=(.*)&export=download$", re.IGNORECASE
+    )
+
+    def __init__(self, url, content_type=None):
+        self.url = url
+
+        if content_type is None and self.GOOGLE_DRIVE_REGEX.match(url):
+            content_type = "pdf"
+
+        self._content_type = content_type
+
+    @property
+    def is_pdf(self):
+        return self._content_type == "pdf"
+
+    @property
+    def is_html(self):
+        return self._content_type == "html"
+
+
+class _ViaClient:
+    """A small wrapper to make calling Via easier."""
+
+    def __init__(self, service_url, legacy_service_url, host_url, legacy_mode=False):
+        self.service_url = urlparse(service_url)
+        self.legacy_service_url = legacy_service_url
+
+        # Default via parameters
+        self.options = {
+            "via.client.openSidebar": "1",
+            "via.client.requestConfigFromFrame.origin": host_url,
+            "via.client.requestConfigFromFrame.ancestorLevel": "2",
+        }
+        self.legacy_mode = legacy_mode
+
+    def url_for(self, doc):
+        if not isinstance(doc, _ViaDoc):
+            raise TypeError(f"Expected doc to be a ViaDoc, not '{type(doc)}'")
+
+        if self.legacy_mode or doc.is_html:
+            return self._legacy_via_url(doc.url)
+
+        if doc.is_pdf:
+            return self._url_for("/pdf", doc.url)
+
+        return self._url_for("/route", doc.url)
+
+    def _url_for(self, path, doc_url):
+        options = {"url": doc_url}
+        options.update(self.options)
+
+        return self.service_url._replace(path=path, query=urlencode(options)).geturl()
```

Comment:
This is actually the second change here (that isn't stylistic) as we needed to be able to construct both `/pdf` and `/route` URLs here.